crates/bevy_render/src/camera/visible_entities.rs

```diff
@@ -24,22 +24,92 @@ impl VisibleEntities {
     }
 }
 
+/// A mask that describes which rendering group an entity belongs to.
+/// Cameras with this component will only render entities with a matching
+/// mask.
+#[derive(Copy, Clone, Debug, Reflect, PartialEq, Eq, PartialOrd, Ord)]
+#[reflect(Component)]
+pub struct RenderingMask(pub u32);
+
+impl Default for RenderingMask {
+    fn default() -> Self {
+        RenderingMask(0)
+    }
+}
+
+impl RenderingMask {
+    pub fn group(n: u8) -> Self {
+        RenderingMask::default().with_group(n)
+    }
+
+    pub fn with_group(mut self, group: u8) -> Self {
+        self.0 = self.0 | (1 << group);
+        self
+    }
+
+    pub fn without_group(mut self, group: u8) -> Self {
+        self.0 = self.0 | (0 << group);
+        self
+    }
+
+    /// Determine if a `RenderingMask` matches.
+    /// `RenderingMask`s match if the first mask contains any of the groups
+    /// in the second, or if both masks are `0`.
+    pub fn matches(&self, other: &RenderingMask) -> bool {
+        ((self.0 & other.0) > 0) || (self.0 == 0 && other.0 == 0)
+    }
+}
+
+#[cfg(test)]
+mod rendering_mask_tests {
+    use super::RenderingMask;
+
+    #[test]
+    fn rendering_mask_sanity() {
+        // groups match groups
+        assert!(RenderingMask::group(1).matches(&RenderingMask::group(1)));
+        // a group of 0 means the mask is just 1 bit
+        assert!(RenderingMask::group(0).matches(&RenderingMask(1)));
+        // a mask will match another mask containing any similar groups
+        assert!(RenderingMask::group(0)
+            .with_group(3)
+            .matches(&RenderingMask::group(3)));
+    }
+}
+
 pub fn visible_entities_system(
-    mut camera_query: Query<(&Camera, &GlobalTransform, &mut VisibleEntities)>,
-    visible_query: Query<(Entity, &Visible)>,
+    mut camera_query: Query<(
+        &Camera,
+        &GlobalTransform,
+        &mut VisibleEntities,
+        Option<&RenderingMask>,
+    )>,
+    visible_query: Query<(Entity, &Visible, Option<&RenderingMask>)>,
     visible_transform_query: Query<&GlobalTransform, With<Visible>>,
 ) {
-    for (camera, camera_global_transform, mut visible_entities) in camera_query.iter_mut() {
+    for (camera, camera_global_transform, mut visible_entities, maybe_camera_mask) in
+        camera_query.iter_mut()
+    {
         visible_entities.value.clear();
         let camera_position = camera_global_transform.translation;
 
         let mut no_transform_order = 0.0;
         let mut transparent_entities = Vec::new();
-        for (entity, visible) in visible_query.iter() {
+        for (entity, visible, maybe_ent_mask) in visible_query.iter() {
             if !visible.is_visible {
                 continue;
             }
 
+            let camera_mask = maybe_camera_mask
+                .map(|m| *m)
+                .unwrap_or_else(|| RenderingMask(0));
```

Comment:
I think using `unwrap_or_default` would be better to define the default value only once